### ISSUE TITLE
Inference in SpeechLM always uses microbatch=1 and tokens_to_generate=64, ignoring custom settings

### ISSUE THREAD

**Issue by hedhoud** [NVIDIA-NeMo/NeMo#14928]
**Describe the bug**

When running inference with [SpeechLM](https://github.com/NVIDIA-NeMo/NeMo/tree/main/examples/speechlm) for evaluation, following the steps in the [inference section](https://github.com/NVIDIA-NeMo/NeMo/tree/main/examples/speechlm#inference) on large datasets, the model always enforces a constant microbatch size of 1 and uses the default tokens_to_generate=64. User-provided configurations are ignored.
This makes it impractical to evaluate on large datasets or modify inference behavior.

**Steps/Code to reproduce bug**
Logs observed:
```text
[NeMo I 2025-10-10 19:40:11 num_microbatches_calculator:228] setting number of microbatches to constant 1
[NeMo W 2025-10-10 19:40:20 speech_to_text_llm_model:897] inference_config is not set. Use default: {'tokens_to_generate': 64}
```
Example command used for training/inference:

```bash 
export GLOBAL_BATCH=3
export MICRO_BATCH=3

srun torchrun $DISTRIBUTED_ARGS $LOCAL_FOLDER/speech_to_text_llm_validate.py \
    --config-path $CONFIG_PATH \
    --config-name $CONFIG_NAME \
    ~data.train_ds \
    data.common.global_batch_size=$GLOBAL_BATCH \
    data.common.micro_batch_size=$MICRO_BATCH \
    data.common.prompt_format='llama3' \
    ++data.validation_ds.input_cfg=$INPUT_CFG_INFER \
    ++data.validation_ds.global_batch_size=$GLOBAL_BATCH \
    ++data.validation_ds.micro_batch_size=$MICRO_BATCH \
    ++data.validation_ds.batch_size=$MICRO_BATCH \
    data.validation_ds.tokens_to_generate=128 \
    data.validation_ds.num_workers=30 \
    data.validation_ds.pin_memory=True \
    data.validation_ds.defer_setup=true \
    data.validation_ds.use_lhotse=true \
    data.validation_ds.metric.name='bleu' \
    ++data.validation_ds.write_predictions_to_file=true \
    ++data.validation_ds.output_dir=$OUTPUT_DIR \
    trainer.limit_val_batches=1\
    ++trainer.val_check_interval=1\
    trainer.devices=1 \
    trainer.num_nodes=1 \
    ++resume.resume_from_path=$CKPT_PATH 
```
**Expected behavior**

Inference should respect user-provided configuration for:
- microbatch_size
- tokens_to_generate

These values should be overrideable via YAML, CLI, or inference arguments instead of being forced to defaults.

**Environment details**
GPU: H100
NeMo version: 2.3.1

**Additional context**
This limitation makes it difficult to evaluate models on large datasets or perform batch-based inference efficiently.

**Comment by stevehuang52** [comment]
Thanks for raising the bug. Could you please try following this [PR](https://github.com/NVIDIA-NeMo/NeMo/pull/14931/files) and see it if fixes the issue? Try puling this branch `heh/speechlm_infer_cfg` and then add these params in inference command:
```bash
data.common.tokens_to_generate=256 \
++model.inference_config.tokens_to_generate=256 \
++model.inference_config.temperature=1.0 \
++model.inference_config.top_k=50 \
++model.inference_config.top_p=0.95 \
++model.inference_config.greedy=false \ 
++model.inference_config.repetition_penalty=1.0 \
```

> [NeMo I 2025-10-10 19:40:11 num_microbatches_calculator:228] setting number of microbatches to constant 1

This message doesn't mean that the micro batch size is set to one, it's more of a gradient accumulation thing.

**Comment by hedhoud** [comment]
Thanks @stevehuang52 for the quick response! I haven't tested the PR yet because my current environment is built with NeMo 2.3.1, and updating it to use that branch is a bit complicated at the moment.

I also ran into another issue with inference: it only outputs the same number of samples as the global batch size, and nothing more. For example, if I set the global batch size to 5, the prediction file contains only 5 samples, even when the dataset is much larger.

I also had a related question: are the ROUGE scores computed over the entire dataset, or only on the first N samples corresponding to the batch size?

**Comment by stevehuang52** [comment]
@hedhoud I've added another small line to potentially fix the issue of missing data samples in output file. I'm not sure why it would be limited to the global batch size, maybe there's something wrong in these [lines](https://github.com/NVIDIA-NeMo/NeMo/blob/main/nemo/collections/speechlm/models/speech_to_text_llm_model.py#L1155-L1158), but I think the new [simple change](https://github.com/NVIDIA-NeMo/NeMo/pull/14931/files#diff-538b313fb72df6b0cb2ac018090e38570459358bd35c295d5744a5a7cb20b11fR137) in this PR should make the samples have different keys when trying to deduplicate.

> are the ROUGE scores computed over the entire dataset, or only on the first N samples corresponding to the batch size?

It's calculated based on the `deduplicated_outputs` obtained in [here](https://github.com/NVIDIA-NeMo/NeMo/blob/main/nemo/collections/speechlm/models/speech_to_text_llm_model.py#L1155-L1158). It would be better to only run on single GPU and add a breakpoint there to see if the number of deduplicated outputs matches with the number of samples in the dataset.

Overall the PR is quite simple, please feel free to just copy and paste the changes to your code and let me know if it works. I think the main issue comes from incorrect number of `deduplicated_outputs`.

**Comment by hedhoud** [comment]
Hi @stevehuang52,

Thanks for your help! The change you suggested fixed the inference configs, but the issue of missing data samples in the output file still persists. I always run inference on a single GPU.

Below is a snippet of how I run it via SLURM:

``` bash

#!/bin/bash

# Using 1 Node and 1 GPU
...

export PYTHONPATH="$PYTHONPATH:/path/to/NeMo"
LOCAL_FOLDER=$(pwd)
TMP_CACHE="$SCRATCH/.cache"
EXPERIMENT_FOLDER="$SCRATCH/experiments"

export TMPDIR=$JOBSCRATCH
export HYDRA_FULL_ERROR=1
export HF_DATASETS_OFFLINE=1
export HF_HUB_OFFLINE=1
export TRANSFORMERS_OFFLINE=1
export NEMO_MODELS_CACHE=$TMP_CACHE
export SPEECH_ENCODER_MODEL="$NEMO_MODELS_CACHE/asr_model/canary-1b-v2.nemo"

# =====================================================
# Parallelism Setup
# =====================================================
TOTAL_GPUS=$(( SLURM_NNODES * SLURM_GPUS_ON_NODE ))

TP=1   # Tensor Parallelism
PP=1   # Pipeline Parallelism
CP=1   # Context Parallelism
DP=$(( TOTAL_GPUS / (TP * PP * CP) ))

export GLOBAL_BATCH=2
export MICRO_BATCH=$(( GLOBAL_BATCH / DP ))
export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK

# =====================================================
# Configs & Paths
# =====================================================
EXPERIMENT_NAME="..."

CONFIG_PATH="$LOCAL_FOLDER/conf"
CONFIG_NAME="config_fake_linear.yaml"
INPUT_CFG_INFER="$LOCAL_FOLDER/conf/input_cfg_infer.yaml"

CKPT_PATH="$EXPERIMENT_FOLDER/$EXPERIMENT_NAME/checkpoints/ckpt_eval/"
OUTPUT_DIR="$EXPERIMENT_FOLDER/$EXPERIMENT_NAME/output"
mkdir -p "$OUTPUT_DIR"

# =====================================================
# Inference Launch
# =====================================================
srun torchrun $DISTRIBUTED_ARGS \
  $LOCAL_FOLDER/speech_to_text_llm_validate.py \
    --config-path $CONFIG_PATH \
    --config-name $CONFIG_NAME \
    ~data.train_ds \
    data.common.global_batch_size=$GLOBAL_BATCH \
    data.common.micro_batch_size=$MICRO_BATCH \
    data.common.prompt_format='llama3' \
    ++data.validation_ds.input_cfg=$INPUT_CFG_INFER \
    ++data.validation_ds.global_batch_size=$GLOBAL_BATCH \
    ++data.validation_ds.micro_batch_size=$MICRO_BATCH \
    data.validation_ds.num_workers=30 \
    data.validation_ds.pin_memory=True \
    data.validation_ds.defer_setup=true \
    data.validation_ds.use_lhotse=true \
    data.validation_ds.metric.name='bleu' \
    ++data.validation_ds.write_predictions_to_file=true \
    ++data.validation_ds.output_dir=$OUTPUT_DIR \
    data.common.tokens_to_generate=256 \
    ++model.inference_config.tokens_to_generate=256 \
    ++model.inference_config.temperature=1.0 \
    ++model.inference_config.top_k=50 \
    ++model.inference_config.top_p=0.95 \
    ++model.inference_config.greedy=false \
    ++model.inference_config.repetition_penalty=1.0 \
    ++trainer.limit_val_batches=1 \
    ++trainer.val_check_interval=1 \
    trainer.devices=-1 \
    trainer.num_nodes=$SLURM_NNODES \
    ++resume.resume_from_path=$CKPT_PATH \
    name="${EXPERIMENT_NAME}_last_eval" \
```


Here’s a snippet from the logs:

```text
[NeMo I 2025-10-30 16:50:56 speech_to_text_llm_model:1275] Setting inference config: {'tokens_to_generate': 256, 'temperature': 1.0, 'top_k': 50, 'top_p': 0.95, 'greedy': False, 'repetition_penalty': 1.0}
[NeMo I 2025-10-30 16:50:56 base:343] Use preset vocab_size: 152064, original vocab_size: 151665, dummy tokens: 399.
...
[NeMo I ... peft:238] Loading adapters from /path/to/checkpoints/weights
[NeMo I ...] Initializing model parallel
..
[NeMo I ... utils:532] Setting up DistributedDataParallel with config DistributedDataParallelConfig(...)
[NeMo I ... pl:305] Global Checkpoint Load : Rank : 0 : Time spent in load_checkpoint: 0.564s
[NeMo W ... peft:277] MegatronOptimizerModule not found in trainer callbacks
[NeMo I ... num_microbatches_calculator:228] setting number of microbatches to constant 1
Validation: iteration 1/1
[NeMo I ... speech_to_text_llm_model:1187] val_bleu_test_multi_turn: 0.124
[NeMo I ... speech_to_text_llm_model:1194] Total deduplicated inference data size: 2 to 2
[NeMo I ... speech_to_text_llm_model:1220] Predictions saved to /path/to/output/speechlm_pred_val_bleu_test_multi_turn_inputs_preds_labels.jsonl

```

Even with these configs, the number of deduplicated outputs seems smaller than the dataset size. I suspect the root cause might still be in the deduplication logic itself rather than batch or GPU setup.

Would you suggest adding more debug logging inside the deduplication step to see why some samples might be dropped?

**Comment by stevehuang52** [comment]
> [NeMo I ... speech_to_text_llm_model:1194] Total deduplicated inference data size: 2 to 2

From this logging message it seems that the deduplication logic didn't filter out anything, as the total num of samples is 2, and after deduplication it remains as 2. 

Ah, the issue might be in your command `++trainer.limit_val_batches=1 \`, you limit the validation to only one batch , that's why it stops after a single batch and you get outputs for 2 samples (global batch=2). Could you please try dropping this line?

**Comment by hedhoud** [comment]
After removing `++trainer.limit_val_batches=1 \`, the validation process enters an infinite loop, it never terminates and no results are produced.
I suspect the issue might be related to the configuration parameter `data.validation_ds.force_iterable_dataset=true`, as I’ve encountered similar reports of infinite validation iterations when this flag is enabled.

Below are some relevant excerpts from the logs:

```text
[NeMo I 2025-10-31 15:49:18 pl:283] Using <megatron.core.dist_checkpointing.strategies.fully_parallel.FullyParallelLoadStrategyWrapper object at 0x146d6269c490> dist-ckpt load strategy.
[NeMo I 2025-10-31 15:49:18 pl:305] Global Checkpoint Load : Rank : 0 : Start time : 1761922158.109s : Time spent in load_checkpoint: 0.764s
[NeMo W 2025-10-31 15:49:18 peft:277] MegatronOptimizerModule not found in trainer callbacks. finalize_model_grads is not properly set up for PEFT.
[NeMo I 2025-10-31 15:49:18 num_microbatches_calculator:228] setting number of microbatches to constant 1
...
Validation: iteration 1/inf
Validation: iteration 2/inf
Validation: iteration 3/inf
...
Validation: iteration 100/inf
...
```

Validation should stop after processing all batches in the validation dataset and produce evaluation metrics or model outputs. However, in this case, it keeps running indefinitely and no results are generated.

**Comment by stevehuang52** [comment]
Yes it could be due to `data.validation_ds.force_iterable_dataset=true`. Could you please try dropping it and also adding `++data.validation_ds.force_finite=true \` and `++model.data.validation_ds.force_map_dataset=true  \`?

**Comment by hedhoud** [comment]
Thanks @stevehuang52, the problem is solved. For the fix, I used `++data.validation_ds.force_map_dataset=true` instead of `++model.data.validation_ds.force_map_dataset=true`

**Comment by stevehuang52** [comment]
Thanks for confirming. I'll update the README.md file to avoid confusions in the future~
